diff --git "a/python/core/auto_generated/auth/qgsauthmanager.sip.in" "b/python/core/auto_generated/auth/qgsauthmanager.sip.in"
index a0a02f71d..faac9c9fe 100644
--- "a/python/core/auto_generated/auth/qgsauthmanager.sip.in"
+++ "b/python/core/auto_generated/auth/qgsauthmanager.sip.in"
@@ -820,7 +820,7 @@ Clear an authentication config from its associated authentication method cache
 
   public:
   protected:
-
+    explicit QgsAuthManager();
 };
 
 /************************************************************************
