diff --git a/src/gui/qgscurveeditorwidget.cpp b/src/gui/qgscurveeditorwidget.cpp
index e825480..38609fc 100644
--- a/src/gui/qgscurveeditorwidget.cpp
+++ b/src/gui/qgscurveeditorwidget.cpp
@@ -34,6 +34,8 @@
 #include <qwt_plot_layout.h>
 #include <qwt_symbol.h>
 #include <qwt_legend.h>
+#include <qwt_scale_div.h>
+#include <qwt_scale_map.h>
 
 #include <qwt_plot_renderer.h>
 #include <qwt_plot_histogram.h>
